### ISSUE TITLE
Use pip freeze to add dependency tracking back to Clipper

### ISSUE THREAD

**Issue by dcrankshaw** [ucbrise/clipper#459]
In #435 we removed a feature that automatically tracked down Python package dependencies and installed them in model containers at runtime. The reason to remove the dependency handling was twofold.
1. The dependency handling relied on Anaconda, and conda was causing a ton of issues in our Docker builds. I decided to remove conda from our Docker containers and just build minimal docker images.
2. The dependency-crawling part of Clipper admin didn't work very well. It used a semi-private conda interface which was very sensitive to the specific conda version people were running on and therefore kept breaking. Furthermore, when we installed these dependencies at runtime it caused container start times to be really high.

@santi81 suggested that instead of removing the dependency management entirely, we could just use pip freeze instead. This is an excellent idea, and allows us to offer some choice over the behavior to users. Specifically, when users deploy a model, we can let them choose between the following behaviors:
1. Disabling dependency management and just doing it manually. This is the only supported behavior after #435 was merged.
2. Tracking dependencies but only installing them at runtime if there was an import error in the container (and thus keeping image sizes small). This was the default behavior before.
3. Tracking dependencies and installing them in the container at build time which will increase the image size but should make container start times fast.

**Comment by simon-mo** [comment]
`pip freeze` outputs whatever user `pip install` before. It can be a huge list if the user is not in a virtualenv/conda env and involves a lot of unrelated packages. Option 3 needs a clever solution that track the exact packages needed by the user's model.

**Comment by dcrankshaw** [comment]
@simon-mo Yeah that's what the [module dependency analyzer](https://github.com/ucbrise/clipper/blob/931561fa02e790af25483455948ff53a3f2e47f2/clipper_admin/clipper_admin/deployers/module_dependency.py) attempts to do. But part of the problem with the just-removed existing implementation was that it captured a bunch of spurious packages that weren't needed.
